<h3 style = 'text-align:center;'>Building a Simple Neural Network to predict the Authenticity of a currency Note.</h3>

<img src = 'Images/neural-network.png'/>

In [31]:
import pandas as pd

In [33]:
banknote = pd.read_csv('banknote.csv')
banknote.sample(4)

variance    skew  curtosis  entropy  class
640    4.16650 -0.4449   0.23448  0.27843      0
1312   0.55939 -0.3104   0.18307  0.44653      1
279    1.92650  7.7557  -0.16823 -3.07710      0
129    3.46630  1.1112   1.74250  1.33880      0

### Importing Libraries and dependencies

In [13]:
import csv
import tensorflow as tf

In [14]:
from sklearn.model_selection import train_test_split

### Read Data into file

In [15]:
with open('banknote.csv') as p:
    reader = csv.reader(p)
    next(reader)
    
    data = []
    for peace in reader:
        data.append({'evidence':[float(cell) for cell in peace[:4]],
                    'label': 1 if peace[4] == '0' else 0})

### Seperating data into training and testing groups

In [16]:
evidence = [peace['evidence'] for peace in data]
label = [peace['label'] for peace in data]
X_train, X_test,y_train,y_test = train_test_split(evidence,label,test_size=0.4)

In [30]:
#evidence

In [29]:
#label

### Create a Neural Network

In [17]:
#Container for layers
model = tf.keras.models.Sequential()

In [18]:
#Add a hidden layer with 8 units, with ReLu activation function.
model.add(tf.keras.layers.Dense(8, input_shape =(4,),activation='relu'))

In [19]:
#Add an output layer with 1 units, with sigmoid activation function.
model.add(tf.keras.layers.Dense(1,activation='sigmoid'))

### Train the Neural Network

In [20]:
model.compile(
    optimizer = 'adam',
    loss = 'binary_crossentropy',
    metrics= ['accuracy']
)
model.fit(X_train,y_train, epochs=20)

#Lets see how well the model performs
model.evaluate(X_test, y_test, verbose = 2)

Epoch 1/20
26/26 [==============================] - 2s 6ms/step - loss: 1.3176 - accuracy: 0.4945
Epoch 2/20
26/26 [==============================] - 0s 4ms/step - loss: 0.9584 - accuracy: 0.5553
Epoch 3/20
26/26 [==============================] - 0s 4ms/step - loss: 0.6865 - accuracy: 0.6756
Epoch 4/20
26/26 [==============================] - 0s 4ms/step - loss: 0.5103 - accuracy: 0.7861
Epoch 5/20
26/26 [==============================] - 0s 4ms/step - loss: 0.4106 - accuracy: 0.8712
Epoch 6/20
26/26 [==============================] - 0s 4ms/step - loss: 0.3579 - accuracy: 0.9052
Epoch 7/20
26/26 [==============================] - 0s 4ms/step - loss: 0.3257 - accuracy: 0.9307
Epoch 8/20
26/26 [==============================] - 0s 4ms/step - loss: 0.3016 - accuracy: 0.9478
Epoch 9/20
26/26 [==============================] - 0s 4ms/step - loss: 0.2819 - accuracy: 0.9526
Epoch 10/20
26/26 [==============================] - 0s 4ms/step - loss: 0.2641 - accuracy: 0.9526
Epoch 11/20
26/26 [

[0.12313269078731537, 0.9908925294876099]

We can see from the results above that the model performs well with an accuracy of 99.09% and loss of 0.123, and we can notice a significant increase in the accuracy of our training process